File: rust-backend/backend/src/handlers/todos.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use uuid::Uuid;

use crate::{
    db, error::AppError, error::Result, models::{CreateTodo, Todo, UpdateTodo}, AppState
};

const CACHE_KEY: &str = "todos:list";
const CACHE_TTL: u64 = 60; // 60 seconds

pub async fn list_todos(State(state): State<AppState>) -> Result<Json<Vec<Todo>>> {
    // Try to get from cache
    if let Ok(Some(cached)) = state.cache.get(CACHE_KEY).await {
        if let Ok(todos) = serde_json::from_str::<Vec<Todo>>(&cached) {
            tracing::debug!("Returning cached todos list");
            return Ok(Json(todos));
        }
    }

    // Cache miss or parse error, fetch from database
    let todos = db::get_all_todos(&state.db).await?;

    // Store in cache
    if let Ok(json) = serde_json::to_string(&todos) {
        let _ = state.cache.set(CACHE_KEY, &json, CACHE_TTL).await;
        tracing::debug!("Cached todos list for {} seconds", CACHE_TTL);
    }

    Ok(Json(todos))
}

pub async fn get_todo(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<Todo>> {
    let todo = db::get_todo_by_id(&state.db, id).await?;
    Ok(Json(todo))
}

pub async fn create_todo(
    State(state): State<AppState>,
    Json(data): Json<CreateTodo>,
) -> Result<(StatusCode, Json<Todo>)> {
    // Validate input
    data.validate()
        .map_err(|e| AppError::BadRequest(e))?;

    let todo = db::create_todo(&state.db, data).await?;

    // Invalidate cache
    let _ = state.cache.delete(CACHE_KEY).await;

    Ok((StatusCode::CREATED, Json(todo)))
}

pub async fn update_todo(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(data): Json<UpdateTodo>,
) -> Result<Json<Todo>> {
    // Validate input
    data.validate()
        .map_err(|e| AppError::BadRequest(e))?;

    let todo = db::update_todo(&state.db, id, data).await?;

    // Invalidate cache
    let _ = state.cache.delete(CACHE_KEY).await;

    Ok(Json(todo))
}

pub async fn delete_todo(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<StatusCode> {
    db::delete_todo(&state.db, id).await?;

    // Invalidate cache
    let _ = state.cache.delete(CACHE_KEY).await;

    Ok(StatusCode::NO_CONTENT)
}

File: go-fullstack/backend/handlers/todos.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/locus/go-fullstack-backend/cache"
	"github.com/locus/go-fullstack-backend/database"
	"github.com/locus/go-fullstack-backend/models"
)

const (
	todoCacheKey    = "todos:list"
	todoCacheTTL    = 60 * time.Second
	defaultTimeout  = 10 * time.Second
)

// ListTodos handles GET /api/todos
func ListTodos(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), defaultTimeout)
	defer cancel()

	// Try to get from cache first
	cached, err := cache.Client.Get(ctx, todoCacheKey).Result()
	if err == nil {
		var todos []models.Todo
		if err := json.Unmarshal([]byte(cached), &todos); err == nil {
			c.JSON(http.StatusOK, gin.H{"data": todos, "cached": true})
			return
		}
	}

	// Cache miss, query database
	query := `SELECT id, title, completed, created_at FROM todos ORDER BY created_at DESC`
	rows, err := database.DB.Query(ctx, query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch todos"})
		return
	}
	defer rows.Close()

	todos := []models.Todo{}
	for rows.Next() {
		var todo models.Todo
		if err := rows.Scan(&todo.ID, &todo.Title, &todo.Completed, &todo.CreatedAt); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to scan todo"})
			return
		}
		todos = append(todos, todo)
	}

	if err := rows.Err(); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Error iterating todos"})
		return
	}

	// Cache the results
	if data, err := json.Marshal(todos); err == nil {
		cache.Client.Set(ctx, todoCacheKey, data, todoCacheTTL)
	}

	c.JSON(http.StatusOK, gin.H{"data": todos, "cached": false})
}

// GetTodo handles GET /api/todos/:id
func GetTodo(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), defaultTimeout)
	defer cancel()

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid todo ID"})
		return
	}

	var todo models.Todo
	query := `SELECT id, title, completed, created_at FROM todos WHERE id = $1`
	err = database.DB.QueryRow(ctx, query, id).Scan(&todo.ID, &todo.Title, &todo.Completed, &todo.CreatedAt)

	if err == pgx.ErrNoRows {
		c.JSON(http.StatusNotFound, gin.H{"error": "Todo not found"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch todo"})
		return
	}

	c.JSON(http.StatusOK, todo)
}

// CreateTodo handles POST /api/todos
func CreateTodo(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), defaultTimeout)
	defer cancel()

	var input models.CreateTodoInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var todo models.Todo
	query := `INSERT INTO todos (title, completed) VALUES ($1, $2) RETURNING id, title, completed, created_at`
	err := database.DB.QueryRow(ctx, query, input.Title, input.Completed).
		Scan(&todo.ID, &todo.Title, &todo.Completed, &todo.CreatedAt)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create todo"})
		return
	}

	// Invalidate cache
	cache.Client.Del(ctx, todoCacheKey)

	c.JSON(http.StatusCreated, todo)
}

// UpdateTodo handles PUT /api/todos/:id
func UpdateTodo(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), defaultTimeout)
	defer cancel()

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid todo ID"})
		return
	}

	var input models.UpdateTodoInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Build dynamic update query
	updates := []string{}
	args := []interface{}{}
	argIndex := 1

	if input.Title != nil {
		updates = append(updates, fmt.Sprintf("title = $%d", argIndex))
		args = append(args, *input.Title)
		argIndex++
	}

	if input.Completed != nil {
		updates = append(updates, fmt.Sprintf("completed = $%d", argIndex))
		args = append(args, *input.Completed)
		argIndex++
	}

	if len(updates) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No fields to update"})
		return
	}

	args = append(args, id)
	query := fmt.Sprintf(`
		UPDATE todos SET %s
		WHERE id = $%d
		RETURNING id, title, completed, created_at
	`,
		string(updates[0]),
		argIndex,
	)

	// Build full query with all updates
	if len(updates) > 1 {
		query = fmt.Sprintf(`
			UPDATE todos SET %s
			WHERE id = $%d
			RETURNING id, title, completed, created_at
		`,
			joinUpdates(updates),
			argIndex,
		)
	}

	var todo models.Todo
	err = database.DB.QueryRow(ctx, query, args...).
		Scan(&todo.ID, &todo.Title, &todo.Completed, &todo.CreatedAt)

	if err == pgx.ErrNoRows {
		c.JSON(http.StatusNotFound, gin.H{"error": "Todo not found"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update todo"})
		return
	}

	// Invalidate cache
	cache.Client.Del(ctx, todoCacheKey)

	c.JSON(http.StatusOK, todo)
}

// DeleteTodo handles DELETE /api/todos/:id
func DeleteTodo(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), defaultTimeout)
	defer cancel()

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid todo ID"})
		return
	}

	query := `DELETE FROM todos WHERE id = $1`
	result, err := database.DB.Exec(ctx, query, id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete todo"})
		return
	}

	if result.RowsAffected() == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Todo not found"})
		return
	}

	// Invalidate cache
	cache.Client.Del(ctx, todoCacheKey)

	c.Status(http.StatusNoContent)
}

// Helper function to join update clauses
func joinUpdates(updates []string) string {
	result := ""
	for i, update := range updates {
		if i > 0 {
			result += ", "
		}
		result += update
	}
	return result
}
